File: Android/_setup/_ProjectName_/src/main/java/com/twist/dataengine/entities/TM_ProductFilter.java
package com.twist.dataengine.entities;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by Twist Mobile on 4/8/2016.
 */
public class TM_ProductFilter {

    public static List<TM_ProductFilter> getAll() {
        return allProductFilters;
    }

    public static void clearAll() {
        allProductFilters.clear();
    }

    private static List<TM_ProductFilter> allProductFilters = new ArrayList<>();

    private TM_ProductFilter() {
        register();
    }

    public static TM_ProductFilter getForCategory(int categoryId) {
        for (TM_ProductFilter filter : allProductFilters) {
            if (filter.categoryId == categoryId)
                return filter;
        }

        TM_ProductFilter productFilter = TM_ProductFilter.getWithCategoryId(categoryId);
        productFilter.maxPrice = Integer.MAX_VALUE;
        productFilter.minPrice = -1;
        productFilter.maxDiscount = Integer.MAX_VALUE;
        productFilter.minDiscount = -1;
        productFilter.register();
        return productFilter;
    }

    public static TM_ProductFilter getWithCategoryId(int categoryId) {
        if (categoryId != -1) {
            for (TM_ProductFilter filter : allProductFilters) {
                if (filter.categoryId == categoryId)
                    return filter;
            }
        }
        TM_ProductFilter filter = new TM_ProductFilter();
        filter.categoryId = categoryId;
        return filter;
    }

    private void register() {
        allProductFilters.add(this);
    }

    public int categoryId = -1;
    public float minPrice;
    public float maxPrice;

    public float minDiscount;
    public float maxDiscount;

    private List<TM_FilterAttribute> attributes = new ArrayList<>();

    public List<TM_FilterAttribute> getAttributes() {
        return attributes;
    }

    public void addAttribute(TM_FilterAttribute attribute) {
        attributes.add(attribute);
    }

    public void clearAttribute() {
        attributes.clear();
    }

    public static boolean attribsLoaded = false;

}

File: IOS/TMStore/TMStore/Classes/Models/TM_ProductFilter.m
//
//  TM_ProductFilter.m
//  TMStore
//
//  Created by Rishabh Jain on 16/05/16.
//  Copyright © 2016 Twist Mobile. All rights reserved.
//

#import "TM_ProductFilter.h"
#import "TM_FilterAttribute.h"
#import "Variables.h"
#import "CategoryInfo.h"
static NSMutableArray* allProductFilters = NULL;

static BOOL attribsLoaded = false;

@interface TM_ProductFilter() {
    NSMutableArray* attributes;//array of filterattributes
}
@end

@implementation TM_ProductFilter
+ (NSMutableArray*)getAll {
    return allProductFilters;
}
+(BOOL)attribsLoaded{
    return attribsLoaded;
}
+(BOOL)attribsLoadedTrue{
    return attribsLoaded = true;
}

- (id)init {
    self = [super init];
    if (self) {
        _categoryId = -1;
       
        attributes = [[NSMutableArray alloc] init];
        self.maxPrice = 100000;
        self.minPrice =  0;
        self.maxDiscount = -1;
        self.minDiscount =  -1;
        [self registered];
    }
    return self;
}
- (void)registered{
    if(allProductFilters == NULL){
        allProductFilters = [[NSMutableArray alloc] init];
    }
    [allProductFilters addObject:self];
}

- (NSMutableArray*)getAttributes {
    return attributes;
}
- (void)addAttribute:(TM_FilterAttribute*)attribute {
    [attributes addObject:attribute];
}
+ (TM_ProductFilter*)getForCategory:(int)categoryId {
    RLOG(@"*************************allProductFilters*************************  %d",(int)[allProductFilters count] );
    for (TM_ProductFilter* filter in allProductFilters) {
//        RLOG(@"getForCategory:filterCategoryId=%d", filter.categoryId);
        if(filter.categoryId == categoryId) {
            RLOG(@"filter.maxPrice  %f",filter.maxPrice);
            return filter;
        }
    }
    TM_ProductFilter *productFilter = [self getWithCategoryId:categoryId];
//    [productFilter registered];
    return productFilter;
}
+ (TM_ProductFilter*)getWithCategoryId:(int)categoryId {
    RLOG(@"*************************allProductFilters*************************  %d",(int)[allProductFilters count] );
    RLOG(@"getWithCategoryId:filterCategoryId=%d", categoryId);
    if(categoryId != -1) {
        CategoryInfo* cInfo = [CategoryInfo getWithId:categoryId];
        if (cInfo && cInfo.pFilter) {
            return cInfo.pFilter;
        }
        //        for (TM_ProductFilter* filter in allProductFilters) {
        //            RLOG(@"getWithCategoryId:filterCategoryId=%d", filter.categoryId);
        //            if(filter.categoryId == categoryId) {
        //                RLOG(@"filter.categoryId   %d ======  categoryId    %d",filter.categoryId,categoryId);
        //                return filter;
        //            }
        //        }
        TM_ProductFilter* filter = [[TM_ProductFilter alloc] init];
        filter.categoryId = categoryId;
        if (cInfo) {
            cInfo.pFilter = filter;
        }
        return filter;
    }
    return nil;
}
/*
+ (TM_ProductFilter*)getWithCategoryId:(int)categoryId {
    RLOG(@"*************************allProductFilters*************************  %d",(int)[allProductFilters count] );
    if(categoryId != -1) {
        for (TM_ProductFilter* filter in allProductFilters) {
            RLOG(@"getWithCategoryId:filterCategoryId=%d", filter.categoryId);
            if(filter.categoryId == categoryId) {
                RLOG(@"filter.categoryId   %d ======  categoryId    %d",filter.categoryId,categoryId);
                return filter;
            }
        }
        TM_ProductFilter* filter = [[TM_ProductFilter alloc] init];
//        filter.maxPrice = 100000;
//        filter.minPrice =  -1;
//        filter.maxDiscount = 100000;
//        filter.minDiscount =  -1;
        filter.categoryId = categoryId;
        return filter;
    }
    return nil;
}
*/
+ (void)resetAttributeLoaded {
    attribsLoaded = false;
}
@end
